src/commcare_cloud/commands/ansible/service.py

```
@@ -1,328 +1,444 @@
+
+    def _run_action_on_hosts(self, action_fn, host_pattern, process_pattern):
+        if host_pattern:
+            self.environment.inventory_manager.subset(host_pattern)
+
+        if process_pattern:
+            assert process_pattern in self.get_managed_services(), (
+                "{} does not match available sub-processes".format(process_pattern)
             )
 
-    def run_ansible_playbook_for_service_group(self, service_group, args, unknown_args):
-        tags = []
-        action = args.action
-        state = self.DESIRED_STATE_FOR_ACTION[action]
-        args.playbook = "service_playbooks/%s.yml" % service_group
-        services = self.services(service_group, args)
-        if args.only:
-            # for options to act on certain services create tags
-            for service in services:
-                if service:
-                    tags.append("%s_%s" % (action, service))
-            if tags:
-                unknown_args.append('--tags=%s' % ','.join(tags), )
-        unknown_args.extend(['--extra-vars', "desired_state=%s desired_action=%s" % (state, action)])
-        return AnsiblePlaybook(self.parser).run(args, unknown_args)
-
-    def run_supervisor_for_service_group(self, service_group, args, unknown_args):
-        exit_code = 0
-        if service_group == "celery":
-            exit_code = self.run_for_celery(service_group, args.action, args, unknown_args)
-        elif service_group in ["webworkers", "formplayer", "touchforms"]:
-            exit_code = self.run_supervisor_action_for_service_group(service_group, args.action, args, unknown_args)
-        return exit_code
-
-    def services(self, service_group, args):
-        if service_group != 'celery' and args.only:
-            return args.only.split(',')
+            run_on = self.get_inventory_group_for_sub_process(process_pattern)
+            hosts = self.environment.inventory_manager.get_hosts(run_on)
+            if not hosts:
+                raise NoHostsMatch
```

Don't clearly understand the intention here? trying to handle this earlier?
Ansible already raises an error if we run something and there are no corresponding hosts to run it on. I have not checked what `supervisorctl` returns in such a case.

---

If there are no hosts then there's no need to call ansible at all. This just fails early and allows us to output a nice error message